pkg/status/workstatus.go

```diff
@@ -0,0 +1,178 @@
+
+import (
+	"context"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	"github.com/kubestellar/kubestellar/pkg/util"
+)
+
+const workStatusIdentificationIndexKey = "workStatusIdentificationIndex"
+
+type workStatus struct {
+	workStatusRef
+	status map[string]interface{}
+}
+
+func (c *Controller) syncWorkStatus(ctx context.Context, ref workStatusRef) error {
+	workStatus := &workStatus{
+		workStatusRef: ref, //readonly
+		status:        nil,
+	}
+
+	obj, err := c.workStatusLister.ByNamespace(ref.wecName).Get(ref.name)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return fmt.Errorf("failed to get workstatus (%v): %w", ref, err)
+		}
+	} else {
+		status, err := util.GetWorkStatusStatus(obj)
+		if err != nil {
+			return fmt.Errorf("failed to get status from workstatus (%v): %w", ref, err)
+		}
+
+		workStatus.status = status // might be nil
+	}
+
+	combinedStatusSet := c.combinedStatusResolver.NoteWorkStatus(workStatus) // nil .status is equivalent to deleted
+	for combinedStatus := range combinedStatusSet {
+		c.workqueue.AddAfter(combinedStatusRef(combinedStatus.ObjectName.AsNamespacedName().String()), queueingDelay)
+	}
+
+	// only process workstatues with the label for single reported status
+	statusLabelVal, ok := obj.(metav1.Object).GetLabels()[util.BindingPolicyLabelSingletonStatusKey]
+	if !ok {
+		return nil
+	}
+
+	// remove the status if singleton status label value is unset
+	if statusLabelVal == util.BindingPolicyLabelSingletonStatusValueUnset {
+		emptyStatus := make(map[string]interface{})
+		return updateObjectStatus(ctx, &ref.sourceObjectIdentifier, emptyStatus,
+			c.listers, c.wdsDynClient)
+	}
+
+	status, err := util.GetWorkStatusStatus(obj)
+	if err != nil {
+		return err
+	}
+
+	if status == nil {
+		// when status is updated, a new event will trigger update
+		return nil
+	}
+
+	c.logger.Info("updating singleton status", "objectIdentifier", ref.sourceObjectIdentifier)
+	return updateObjectStatus(ctx, &ref.sourceObjectIdentifier, status, c.listers, c.wdsDynClient)
+}
+
+func updateObjectStatus(ctx context.Context, objectIdentifier *util.ObjectIdentifier, status map[string]interface{},
+	listers util.ConcurrentMap[schema.GroupVersionResource, cache.GenericLister], wdsDynClient dynamic.Interface) error {
+	logger := klog.FromContext(ctx)
+
+	gvr := objectIdentifier.GVR()
+	lister, found := listers.Get(gvr)
+	if !found {
+		logger.V(5).Info("could not find lister for gvr", "gvr", gvr)
+		return nil
+	}
+
+	obj, err := lister.ByNamespace(objectIdentifier.ObjectName.Namespace).Get(objectIdentifier.ObjectName.Name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.V(5).Info("could not find object", "objectIdentifier", objectIdentifier)
```

Comment:
(A) A log message should be capitalized like an English sentence.

(B) It would be better if this one were a little more informative, saying that an update is being abandoned because the object is not in the local cache.